src/jacobians.jl

```diff
@@ -0,0 +1,233 @@
+
+"""
+    allocate_jacobian(
+        M_domain::AbstractManifold,
+        M_codomain::AbstractManifold,
+        f,
+        p;
+        basis_domain::AbstractBasis = DefaultOrthonormalBasis(),
+        basis_codomain::AbstractBasis = DefaultOrthonormalBasis(),
+    )
+
+Allocate Jacobian of function `f` with given domain and codomain at point `p`.
+`basis_domain` and `basis_codomain` denote bases of tangent spaces at, respectively, `p` and `f(p)`.
+"""
+function allocate_jacobian(
+    M_domain::AbstractManifold,
+    M_codomain::AbstractManifold,
+    f,
+    p;
+    basis_domain::AbstractBasis = DefaultOrthonormalBasis(),
+    basis_codomain::AbstractBasis = DefaultOrthonormalBasis(),
+)
+    n = number_of_coordinates(M_domain, basis_domain)
+    m = number_of_coordinates(M_codomain, basis_codomain)
+    J = zeros(float(number_eltype(p)), n, m)
+    return J
+end
+
+function jacobian_exp_argument!(
+    M::AbstractManifold,
+    J::AbstractMatrix,
+    p,
+    X,
+    basis_domain::AbstractBasis = DefaultOrthonormalBasis(),
+    basis_codomain::AbstractBasis = DefaultOrthonormalBasis(),
+)
+    Bb = get_basis(M, p, basis_domain)
+    Vs = get_vectors(M, p, Bb)
+    Z = similar(X)
+    p_exp = exp(M, p, X)
+    for i in 1:length(Vs)
+        differential_exp_argument!(M, Z, p, X, Vs[i])
+        get_coordinates!(M, view(J, :, i), p_exp, Z, basis_codomain)
+    end
+    return J
+end
+
+@doc raw"""
+    jacobian_exp_argument(
+        M::AbstractManifold,
+        p,
+        X,
+        basis_domain::AbstractBasis=DefaultOrthonormalBasis(),
+        basis_codomain::AbstractBasis=DefaultOrthonormalBasis(),
+    )
+
+Compute Jacobian of the exponential map with respect to its argument (tangent vector).
+Differential of the exponential map is here considered as a function from ``T_p \mathcal{M}``
+to ``T_{\exp_p X} \mathcal{M}``. Jacobian coefficients are represented in basis `basis_domain`
+in the domain and in `basis_codomain` in the codomain.
+"""
+function jacobian_exp_argument(
+    M::AbstractManifold,
+    p,
+    X,
+    basis_domain::AbstractBasis = DefaultOrthonormalBasis(),
+    basis_codomain::AbstractBasis = DefaultOrthonormalBasis(),
+)
+    J = allocate_jacobian(
+        M,
+        M,
+        jacobian_exp_argument,
+        X;
+        basis_domain = basis_domain,
+        basis_codomain = basis_codomain,
+    )
+    jacobian_exp_argument!(M, J, p, X, basis_domain, basis_codomain)
+    return J
+end
+
+function jacobian_exp_basepoint!(
+    M::AbstractManifold,
+    J::AbstractMatrix,
+    p,
+    X,
+    basis_domain::AbstractBasis = DefaultOrthonormalBasis(),
+    basis_codomain::AbstractBasis = DefaultOrthonormalBasis(),
+)
+    Bb = get_basis(M, p, basis_domain)
+    Vs = get_vectors(M, p, Bb)
+    Z = similar(X)
+    p_exp = exp(M, p, X)
+    for i in 1:length(Vs)
+        differential_exp_basepoint!(M, Z, p, X, Vs[i])
```

Comment:
One could even write this function for a general `Df!` differential instead of implementing this 4 times here? Then each of these functions would be one call to this function, maybe like

`differential_to_jacobian!(M, J, differential_exp_basepoint!, p, X, basis_domain, basis_codomain)`

?

Reply:
hm to me they look all the same with one function to pass, but you see that better than me probably.

Reply:
Ah you are right, my idea does not directly work but something like

`differential_to_jacobian(M_domain, M_range, f, Df, p; bases...)`

should work, when `f = (M,p) -> ...` and `Df = (M, p, X) ->` and instead of `f` one could also provide a signature that works with just `q=f(M,p)` since I think that is the only thin one would need f for.

I am not sure this is beneficial, but it would be the general fallback one could implement.

Reply:
Yes, it might work, though I don't see it being sufficiently useful to rework this PR.

Reply:
Ok, fair enough.